### PR TITLE
Revive PR: Adds mechanism to prevent spam from unhealthy notifications

### PR DESCRIPTION
**Pull request by cieciurm**
**What this PR does / why we need it**:

This PR revives #592
Original description:
> Adds mechanism to prevent spam from unhealthy notifications.

**Which issue(s) this PR fixes**:

None

**Special notes for your reviewer**:

- Merged with `master`
- Tests fixed
- Formatting fixed
- Public API approved

**Does this PR introduce a user-facing change?**:

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [x] End-to-end tests passing
- [x] Extended the documentation
- [x] Provided sample for the feature

### CODE REVIEW

File: src/HealthChecks.UI/Core/HostedService/HealthCheckReportCollector.cs [hunk 1]
```diff
@@ -115,15 +123,15 @@ private async Task<UIHealthReport> GetHealthReportAsync(HealthCheckConfiguration
 
                         using var requestMessage = new HttpRequestMessage(HttpMethod.Get, absoluteUri);
                         requestMessage.Headers.Authorization = new BasicAuthenticationHeaderValue(userInfoArr[0], userInfoArr[1]);
-                        response = await _httpClient.SendAsync(requestMessage, HttpCompletionOption.ResponseHeadersRead);
+                        response = await _httpClient.SendAsync(requestMessage, HttpCompletionOption.ResponseHeadersRead).ConfigureAwait(false);
```
Comment (sungam3r) [inline src/HealthChecks.UI/Core/HostedService/HealthCheckReportCollector.cs:126]: Please remove all added `.ConfigureAwait(false)` to simplify review. I did not add them intentionally across UI projects.

File: test/HealthChecks.UI.Tests/Functional/DatabaseProviders/SqlServerStorageProviderTests.cs [hunk 2]
```diff
@@ -17,7 +17,8 @@ public void register_healthchecksdb_context_with_migrations()
                 {
                     services.AddHealthChecksUI()
                     .AddSqlServerStorage("connectionString", opt => customOptionsInvoked = true);
-                });
+                })
+                .Configure(_ => { });
```
Comment (sungam3r) [inline test/HealthChecks.UI.Tests/Functional/DatabaseProviders/SqlServerStorageProviderTests.cs:21]: Please revert, fixed in #1879 .

File: src/HealthChecks.UI/Core/HostedService/HealthCheckReportCollector.cs (hunk 1 above)
Comment (cieciurm) [reply src/HealthChecks.UI/Core/HostedService/HealthCheckReportCollector.cs]: It's been fixed :)

File: test/HealthChecks.UI.Tests/Functional/DatabaseProviders/SqlServerStorageProviderTests.cs (hunk 2 above)
Comment (cieciurm) [reply test/HealthChecks.UI.Tests/Functional/DatabaseProviders/SqlServerStorageProviderTests.cs]: Done, thx.

**Comment by cieciurm** [comment]
PR adjusted
cc @sungam3r

File: README.md [hunk 3]
```diff
@@ -580,6 +580,19 @@ Webhooks can be configured with configuration providers and also by code. Using
...
 
+**Avoid Fail notification spam**
+
+To prevent you from receiving several failure notifications from your application, a configuration was created to meet this scenario.
+
+```csharp
+
+services.AddHealthChecksUI(setup =>
+{
+    setup.NotifyUnHealthyOneTimeUntilChangeToAvoidNotificationSpam(); //You will only receive one failure notification until the status changes.
```
Comment (sungam3r) [inline README.md:591]:
```suggestion
    setup.NotifyUnHealthyOneTimeUntilChangeToAvoidNotificationSpam(); // You will only receive one failure notification until the status changes.
```

File: src/HealthChecks.UI/Configuration/Settings.cs [hunk 4]
```diff
@@ -51,6 +52,12 @@ public Settings DisableDatabaseMigrations()
             return this;
         }
 
+        public Settings NotifyUnHealthyOneTimeUntilChangeToAvoidNotificationSpam()
```
Comment (sungam3r) [inline src/HealthChecks.UI/Configuration/Settings.cs:55]:
```suggestion
        public Settings NotifyUnHealthyOneTimeUntilChange()
```

?

File: README.md [hunk 5]
```diff
@@ -580,6 +580,19 @@ Webhooks can be configured with configuration providers and also by code. Using
...
 
+**Avoid Fail notification spam**
+
+To prevent you from receiving several failure notifications from your application, a configuration was created to meet this scenario.
+
+```csharp
+
+services.AddHealthChecksUI(setup =>
+{
+    setup.NotifyUnHealthyOneTimeUntilChangeToAvoidNotificationSpam(); // You will only receive one failure notification until the status changes.
```
Comment (sungam3r) [inline README.md:591]:
```suggestion
    setup.NotifyUnHealthyOneTimeUntilChange(); // You will only receive one failure notification until the status changes.
```

File: src/HealthChecks.UI/Core/HostedService/HealthCheckReportCollector.cs [hunk 6]
```diff
@@ -159,7 +167,7 @@ private Uri GetEndpointUri(HealthCheckConfiguration configuration)
 
         private async Task<bool> HasLivenessRecoveredFromFailureAsync(HealthCheckConfiguration configuration)
         {
-            var previous = await GetHealthCheckExecutionAsync(configuration);
+            var previous = await GetHealthCheckExecutionAsync(configuration).ConfigureAwait(false);
```
Comment (sungam3r) [inline src/HealthChecks.UI/Core/HostedService/HealthCheckReportCollector.cs:170]:
```suggestion
            var previous = await GetHealthCheckExecutionAsync(configuration);
```

File: src/HealthChecks.UI/Core/HostedService/HealthCheckReportCollector.cs [hunk 7]
```diff
@@ -173,6 +181,31 @@ private async Task<bool> HasLivenessRecoveredFromFailureAsync(HealthCheckConfigu
...
+               .Where(lf => string.Equals(lf.HealthCheckName, healthCheckName, StringComparison.OrdinalIgnoreCase))
+               .OrderByDescending(lf => lf.LastNotified)
+               .Take(2).ToListAsync();
+
+            if (lastNotifications == null)
+            {
+                return true;
+            }
+
+            if (lastNotifications.Count == 2)
```
Comment (sungam3r) [inline src/HealthChecks.UI/Core/HostedService/HealthCheckReportCollector.cs:196]:
```suggestion
            if (lastNotifications?.Count == 2)
```

File: test/HealthChecks.UI.Tests/HealthChecks.UI.approved.txt [hunk 8]
```diff
@@ -29,6 +29,7 @@ namespace HealthChecks.UI.Configuration
         public HealthChecks.UI.Configuration.Settings ConfigureWebhooksEndpointHttpclient(System.Action<System.IServiceProvider, System.Net.Http.HttpClient> webhooksEndpointHttpClientconfig) { }
         public HealthChecks.UI.Configuration.Settings DisableDatabaseMigrations() { }
         public HealthChecks.UI.Configuration.Settings MaximumHistoryEntriesPerEndpoint(int maxValue) { }
+        public HealthChecks.UI.Configuration.Settings NotifyUnHealthyOneTimeUntilChangeToAvoidNotificationSpam() { }
```
Comment (sungam3r) [inline test/HealthChecks.UI.Tests/HealthChecks.UI.approved.txt:32]:
```suggestion
        public HealthChecks.UI.Configuration.Settings NotifyUnHealthyOneTimeUntilChange() { }
```

**Comment by cieciurm** [comment]
@sungam3r - you suggestion are applied now :)

File: src/HealthChecks.UI/Configuration/Settings.cs [hunk 9]
```diff
@@ -1,143 +1,150 @@
...
+            return this;
+        }
+
+        public Settings DisableDatabaseMigrations()
+        {
+            DisableMigrations = true;
+            return this;
+        }
+
+        public Settings AddNotifyUnHealthyOneTimeUntilChange()
```
Comment (sungam3r) [inline src/HealthChecks.UI/Configuration/Settings.cs:55]:
```suggestion
        public Settings SetNotifyUnHealthyOneTimeUntilChange()
```

File: test/HealthChecks.UI.Tests/HealthChecks.UI.approved.txt [hunk 10]
```diff
@@ -24,6 +24,7 @@ namespace HealthChecks.UI.Configuration
     {
         public Settings() { }
         public HealthChecks.UI.Configuration.Settings AddHealthCheckEndpoint(string name, string uri) { }
+        public HealthChecks.UI.Configuration.Settings AddNotifyUnHealthyOneTimeUntilChange() { }
```
Comment (sungam3r) [inline test/HealthChecks.UI.Tests/HealthChecks.UI.approved.txt:27]:
```suggestion
        public HealthChecks.UI.Configuration.Settings SetNotifyUnHealthyOneTimeUntilChange() { }
```

File: test/HealthChecks.UI.Tests/HealthChecks.UI.approved.txt [hunk 11]
```diff
@@ -24,6 +24,7 @@ namespace HealthChecks.UI.Configuration
     {
         public Settings() { }
         public HealthChecks.UI.Configuration.Settings AddHealthCheckEndpoint(string name, string uri) { }
+        public HealthChecks.UI.Configuration.Settings SetNotifyUnHealthyOneTimeUntilChange() { }
```
Comment (sungam3r) [inline test/HealthChecks.UI.Tests/HealthChecks.UI.approved.txt:27]:
```suggestion
```

Comment (sungam3r) [inline test/HealthChecks.UI.Tests/HealthChecks.UI.approved.txt:27]: Please move after SetMinimumSecondsBetweenFailureNotifications

Comment (cieciurm) [reply test/HealthChecks.UI.Tests/HealthChecks.UI.approved.txt]: Thanks, it's been moved to be sorted alphabetically.

**Comment by cieciurm** [comment]
Public API has been sorted alphabetically and workflow succeeded 🎉

**Pull request merged**